cmd/demo/node_setup.go

```diff
@@ -103,63 +105,125 @@ func (n *node) setup() error {
-		adjAddr, err = deployAdjudicator(n.cb)
-		if err != nil {
-			return
-		}
-		fmt.Println("💭 Adjudicator contract deployed")
-	} else {
-		tmpAdj, err := strToAddress(config.Chain.Adjudicator)
-		if err != nil {
-			return err
-		}
-		adjAddr = ewallet.AsEthAddr(tmpAdj)
+func (n *node) validateAdjudicatorContract(adjAddr common.Address) error {
+	ctx, cancel := getTxContext()
+	defer cancel()
+	return echannel.ValidateAdjudicator(ctx, n.cb, adjAddr)
+}
 
-		ctx, cancel := context.WithTimeout(context.Background(), config.Chain.TxTimeout)
-		defer cancel()
-		if err := echannel.ValidateAdjudicator(ctx, n.cb, adjAddr); err != nil {
-			return errors.WithMessage(err, "validating adjudicator contract")
-		}
+func (n *node) validateAssetHolderContract(assAddr common.Address, adjAddr common.Address) error {
+	ctx, cancel := getTxContext()
+	defer cancel()
+	return echannel.ValidateAssetHolderETH(ctx, n.cb, assAddr, adjAddr)
+}
+
+func stringToAddress(s string) (common.Address, error) {
+	var err error
+	var walletAddr wallet.Address
 
-		fmt.Println("💭 Adjudicator contract validated")
+	if walletAddr, err = strToAddress(s); err != nil {
+		return common.Address{}, err
 	}
 
-	if config.Chain.Assetholder == "deploy" {
-		assAddr, err = deployAsset(n.cb, adjAddr)
-		if err != nil {
-			return
-		}
-		fmt.Println("💭 Asset holder contract deployed")
-	} else {
-		tmpAsset, err := strToAddress(config.Chain.Assetholder)
-		if err != nil {
-			return err
+	return ewallet.AsEthAddr(walletAddr), nil
+}
+
+func (n *node) setupAdjudicatorContract() error {
+	// config options: "[onChainAddress]", "deploy", "[onChainAddress]|deploy"
+
+	tokens := strings.Split(config.Chain.Adjudicator, "|")
+
+	for _, token := range tokens {
+		if token == "deploy" {
+			// deploy contract
+
+			if adjAddr, err := deployAdjudicator(n.cb); err == nil {
+				n.adjAddr = adjAddr
+				n.adjudicatorSetup = true
+				fmt.Printf("🌐 Adjudicator contract deployed at %x\n", adjAddr)
+				return nil
+			}
+
+			fmt.Println("❌ Adjudicator contract deployment failed")
+		} else {
+			// validate contract
+
+			if adjAddr, err := stringToAddress(token); err == nil {
+				if err := n.validateAdjudicatorContract(adjAddr); err == nil {
+					n.adjAddr = adjAddr
+					n.adjudicatorSetup = true
+					fmt.Printf("✅ Adjudicator contract validated at %x\n", adjAddr)
+					return nil
+				}
+			}
+
+			fmt.Printf("❌ Adjudicator contract validation for %v failed\n", token)
 		}
-		assAddr = ewallet.AsEthAddr(tmpAsset)
+	}
 
-		ctx, cancel := context.WithTimeout(context.Background(), config.Chain.TxTimeout)
-		defer cancel()
-		if err := echannel.ValidateAssetHolderETH(ctx, n.cb, assAddr, adjAddr); err != nil {
-			return errors.WithMessage(err, "validating asset holder contract")
+	return errors.New("setup adjudicator contract")
+}
+
+func (n *node) setupAssetHolderContract() error {
+	// config options: "[onChainAddress]", "deploy", "[onChainAddress]|deploy"
+
+	if !n.adjudicatorSetup {
```

Comment:
I think this kind of defensive programming makes it harder to read and does not add real value. Since this is an internal function, the usage can be assumed to be correct.

Reply:
i removed the variable in the most recent version. (it became obsolete with the recent code revision. the point before was that if one does not know how to use the code correctly, the code would check that you cannot setup the asset holder without setting up the adjudicator.)